sycl/include/CL/sycl/handler.hpp

```diff
@@ -792,64 +671,82 @@ class handler {
 #endif // __SYCL_DEVICE_ONLY__
   }
 
-  // single_task version with a kernel represented as a sycl::kernel.
-  // The kernel invocation method has no functors and cannot be called on host.
+  /// Invokes a SYCL kernel.
+  ///
+  /// Executes exactly once. The kernel invocation method has no functors and
+  /// cannot be called on host.
+  ///
+  /// @param SyclKernel is a SYCL kernel object.
   void single_task(kernel SyclKernel) {
     throwIfActionIsCreated();
     verifySyclKernelInvoc(SyclKernel);
     MNDRDesc.set(range<1>{1});
-    MSyclKernel = detail::getSyclObjImpl(std::move(SyclKernel));
+    MSyclKernel = unique_ptr_class<kernel>(new kernel(std::move(SyclKernel)));
     MCGType = detail::CG::KERNEL;
     extractArgsAndReqs();
   }
 
-  // parallel_for version with a kernel represented as a sycl::kernel + range
-  // that specifies global size only. The kernel invocation method has no
-  // functors and cannot be called on host.
+  /// Defines and invokes a SYCL kernel function for the specified range.
+  ///
+  /// The SYCL kernel function is defined as SYCL kernel object. The kernel
+  /// invocation method has no functors and cannot be called on host.
+  ///
+  /// @param NumWorkItems is a range defining indexing space.
+  /// @param WorkItemOffset is an offset to be applied to each work item index.
+  /// @param KernelFunc is a SYCL kernel function.
   template <int Dims>
   void parallel_for(range<Dims> NumWorkItems, kernel SyclKernel) {
     throwIfActionIsCreated();
     verifySyclKernelInvoc(SyclKernel);
-    MSyclKernel = detail::getSyclObjImpl(std::move(SyclKernel));
+    MSyclKernel = unique_ptr_class<kernel>(new kernel(std::move(SyclKernel)));
     MNDRDesc.set(std::move(NumWorkItems));
     MCGType = detail::CG::KERNEL;
     extractArgsAndReqs();
   }
 
-  // parallel_for version with a kernel represented as a sycl::kernel + range
-  // and offset that specify global size and global offset correspondingly.
-  // The kernel invocation method has no functors and cannot be called on host.
+  /// Defines and invokes a SYCL kernel function for the specified range and
+  /// offsets.
+  ///
+  /// The SYCL kernel function is defined as SYCL kernel object.
+  ///
+  /// @param NumWorkItems is a range defining indexing space.
+  /// @param WorkItemOffset is an offset to be applied to each work item index.
+  /// @param KernelFunc is a SYCL kernel function.
   template <int Dims>
-  void parallel_for(range<Dims> NumWorkItems, id<Dims> workItemOffset,
+  void parallel_for(range<Dims> NumWorkItems, id<Dims> WorkItemOffset,
                     kernel SyclKernel) {
     throwIfActionIsCreated();
     verifySyclKernelInvoc(SyclKernel);
-    MSyclKernel = detail::getSyclObjImpl(std::move(SyclKernel));
-    MNDRDesc.set(std::move(NumWorkItems), std::move(workItemOffset));
+    MSyclKernel = unique_ptr_class<kernel>(new kernel(std::move(SyclKernel)));
```

Comment:
Same question here.
What's so bad with getSyclObjImpl?

Reply:
It requires you to know kernel_impl type. This implies linkage of user app against our internal symbols, which leads to broken ABI almost after every second commit.